text/3680-use.md

```diff
@@ -0,0 +1,463 @@
+introducing new bindings for closures and async blocks.
+
+Rather than specifically supporting *lightweight* clones, we could add a syntax
+for closures and async blocks to perform *any* clones (e.g. `async clone` /
+`clone ||`). This would additionally allow expensive clones (such as
+`String`/`Vec`). However, we've had many requests to distinguish between
+expensive and lightweight clones, as well as ecosystem conventions attempting
+to make such distinctions (e.g. past guidance to write `Arc::clone`/`Rc::clone`
+explicitly). Having a syntax that only permits lightweight clones would allow
+users to confidently use that syntax without worrying about an unexpectedly
+expensive operation. We can then provide ways to perform the expensive clones
+explicitly, such as the `use(x = x.clone())` syntax suggested in
+[future possibilities][future-possibilities].
+
+There are myriad names we could use for the trait, invocation syntax, and
+closure / async block syntax. An ideal name for this needs to convey the
+semantic of taking an additional reference to an object, without copying or
+duplicating the object. Other names proposed for this mechanism include
+`Claim`.
+
+We could use a different keyword, if one fits. For instance, we could use `ref`
+rather than `use`. However, this 1) might over-fit to the concept of
+reference-counting specifically, rather than lightweight clones, 2) might
+confuse users who expect it to produce a `&` reference, and 3) for
+closures/blocks, would not generalize to explicit capture lists.
+
+We could use a name that directly references referencing counting; for
+instance, `AddRef`. However, that would be confusing for applications using
+this with objects managed by something other than reference counting, such as
+RCU or hazard pointers, or with (small) objects being copied.
+
+Rather than having the method `Clone::clone`, we could have a method
+`Use::do_use` and have the compiler call that. We could prevent users from
+overriding that method implementation, so that it always calls `Clone::clone`;
+however, having a separate `do_use` method would allow users to call
+`.do_use()` (when wanting to avoid the elision of `.use`) while knowing they
+can't invoke an expensive clone operation. This does not seem worth the
+additional complexity, since most users should invoke `.use` directly.
+
+Rather than using the special syntax `.use`, we could use an ordinary trait
+method and invoke that method directly (e.g. `.claim()`). The syntax provided
+for closures and async blocks could likewise always invoke that method. This
+would not be compatible with adding smarter semantics such as eliding uses,
+however. In addition, `.use` can be used in all editions of Rust (because `use`
+is a reserved keyword), while a new trait with a method like `.claim()` would
+require an import in existing editions and could only become part of the
+prelude in a future edition.
```

Comment:
"This would not be compatible with adding smarter semantics..." is this accurate? Two paragraphs down contradicts it with "We could use an ordinary trait method *and* add a new set of semantics attached to that trait method...".

Imo a trait method would be preferred because there is nothing new to teach; this would just be another trait that happens to have a bit of magic, pretty similar to [`Deref`](https://doc.rust-lang.org/std/ops/trait.Deref.html). Another use of `use` seems more likely to confuse users - it doesn't really give much intuition here.

`.await` is given as a comparison but that is an easily-searchable keyword that has a similar meaning across many languages.

Reply:
I would be fairly strongly opposed to letting the compiler elide explicit calls to a trait method, like `foo.claim()`. We don't have any precedent for that, and I think it is extremely confusing -- it basically changes method call syntax into "maybe call this method, maybe not" syntax. Note that `x.deref()` will always call `deref` so that is not a comparable piece of magic. The magic in `deref` is that new calls to it can be *added* implicitly.